### PR TITLE
Optimize multi_tensor_apply (take 2)

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #119764

### Take 2

The first take (#119153) landed but was reverted because it broke cuda graph for `multi_tensor_apply`. This PR is a reland of #119153:
- Incorporate #119652 so that the optimization can be applied (1) without increasing binary size (2) to all 3 MTA variants without much code duplication.
- Ensure the optimization is compatible with cuda graph.

### Summary

Due to the dynamic nature of the workload, the kernel arguments aren't guaranteed to fit in the static 4kb kernel argument memory. Previously with the apex implementation, we overcame this limitation by dividing a multi_tensor_apply workload into multiple kernel launches. However, this led to low sustained occupancy, affecting the performance of memory bound ops.

Based on the observation that the kernel argument memory limitation doesn't correlate well with available SM resources, we adopt a different approach:
- When the kernel arguments fit into the static kernel argument memory, we use this memory to transfer the arguments.
- Conversely, when the kernel arguments don't fit into the static kernel argument memory, instead of sacrificing sustained occupancy, we use a page-locked cudaMemcpyAsync to transfer the arguments, then perform the entire workload in a single kernel.

This PR only covers `multi_tensor_apply` for tensors. The change can be easily applied to `multi_tensor_apply` for tensors + scalars and `multi_tensor_apply_for_fused_optimizer`.

### Benchmark (WIP)

The only benchmark I've conducted so far on `_foreach_copy_` on a set of sizes that resembles internal workload. I need to benchmarks on more problem sizes. The speedup should vary among problem sizes. **However, I believe this PR should not be slower than the previous impl on any problem sizes.**

The benchmark can be reproduced with [this script](https://gist.github.com/yifuwang/178c1f4bf951c5794ea79c04d90e44fa).

**Baseline**

A single iteration in trace:
image

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_5a59145f-567b-472f-8eef-c61c388d45b4.json
device ms: 1.111, cpu ms: 7.151
memory bandwidth: 1169.825 GB/s
```

**This PR**

A single iteration in trace:
image

```
https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_da060725-62a8-466e-b570-2ad67ff0e29d.json
device ms: 0.892, cpu ms: 0.810
memory bandwidth: 1456.744 GB/s
```

### CODE REVIEW

**Review by eellison — COMMENTED** [review]
Still looking at this but would you add a test?

File: aten/src/ATen/cuda/CUDAGraph.h [hunk 1]
```diff
@@ -85,8 +91,11 @@ struct TORCH_CUDA_CPP_API CUDAGraph {
   // RNG state trackers
   at::Tensor seed_extragraph_;
   at::Tensor offset_extragraph_;
+  std::vector<void*> const_buffers_;
```
Comment (albanD) [inline aten/src/ATen/cuda/CUDAGraph.h:94]: cc @eellison is that an ok thing to do? I know you worked pretty hard to remove all allocations here including workspaces

Comment (eqy) [reply aten/src/ATen/cuda/CUDAGraph.h:94]: Is this something that we want to have generically in CUDA graphs or would it make more sense to have it as something that `MultiTensorApply` holds on to if it is to be used in a graph capture?
I guess I'm missing some context on why this allocation has to be invoked by the graph itself.

**Comment by eellison** [comment]
> For the cuda graph part I plan to add tests if you feel the general direction is reasonable

 even with a different direction we're gonna want tests too

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 2]
```diff
@@ -38,13 +48,178 @@ __device__ __forceinline__ void load_store(
...
+  for (const auto i : c10::irange(ptrs.size())) {
+    pack.offsets[i] = offsets[i];
+    memcpy(buf_tensor.data_ptr<uint8_t>() + offsets[i], ptrs[i], sizes[i]);
+  }
+
+  if (is_capturing) {
+    // When capturing, use a constant buffer to hold the packed vectors
+    at::cuda::get_current_capturing_graph()->alloc_const_buffer(
+        reinterpret_cast<void**>(&pack.buffer_ptr), total_bytes);
+    C10_CUDA_CHECK(cudaMemcpy(
```
Comment (eellison) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:176]: Sorry not totally following - why would this not work with an allocation from the cuda caching allocator ?

File: aten/src/ATen/cuda/CUDAGraph.h (hunk 1 above)
Comment (eellison) [reply aten/src/ATen/cuda/CUDAGraph.h:94]: > I know you worked pretty hard to remove all allocations here including workspaces

The particular problem I had was for allocations which were made with the cuda caching allocator and which would persist outside of the scope of reinvoking the cuda graph. E.g., during capture adding a tensor to cublas workspace which would could then get potentially used in egaer.

 So that doesn't apply here right ?

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 2 above)
Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:176]: Hmm.. now that you mention it, I may have prematurely assumed that it wouldn't work.

By using `CUDACachingAllocator::raw_alloc`, I reserve a buffer from the graph's private pool. By allowing it to "leak", the buffer won't be reassigned for any other purpose, and its lifetime is tied to the `CUDAGraph` object. This seems to be exactly the behavior I want, i.e., a buffer allocated at capture time that persists throughout the lifetime of the `CUDAGraph` object.

Is my understanding correct?

File: aten/src/ATen/cuda/CUDAGraph.h (hunk 1 above)
Comment (yifuwang) [reply aten/src/ATen/cuda/CUDAGraph.h:94]: > Is this something that we want to have generically in CUDA graphs

Hey @eqy, this is not a goal (it might even be something we want to avoid). After seeing @eellison's comment, I think I may have been overcomplicating things. The goal is to allocate a buffer and tie its lifetime to the capturing `CUDAGraph` object, and it's probably achievable by simply allocating it with `CUDACachingAllocator::raw_alloc` and letting it "leak".

**Review by albanD — COMMENTED** [review]
Thanks for the update @yifuwang !

A meta-comment on this PR: if you are trying to speedup the non-cudagraph case mainly and you want to get that done relatively soon, I would suggest to fallback to the old implementation (that launches many kernels) when you detect that cudagraph is being used.

Introducing a concept of "constant buffer" to cudagraph is going to be a significantly trickier change that I expect is going to take a bit of work to get right.
The code as it is today is wrong on a couple of levels: the memory will be re-used by other objects within the current cudagraph (since you Tensor object goes out of scope after the kernel ran) and so you cannot rely on the data to be preserved from one run to the next. Moreover, the cudaMemCopy is very shady here, I'm surprised it is actually not raising an error when being captured. It most likely burns in the CPU side pointer and will re-run the copy (from a CPU Tensor that doesn't exist anymore!).
So going down that path would actually require creating an "immortal" Tensor to preserve both its memory and content forever. Not sure what will be the consequences on the current cudagraph design though.

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 2 above)
Comment (eellison) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:182]: > the buffer won't be reassigned for any other purpose, and its lifetime is tied to the CUDAGraph object. This seems to be exactly the behavior I want, i.e., a buffer allocated at capture time that persists throughout the lifetime of the CUDAGraph object

Right, this already the semantics of when you allocate from the cuda caching allocator during cuda graph capture

**Comment by yifuwang** [comment]
@albanD Thank you for your review and for summarizing the issues. I agree with most of your observations. However, I believe these are implementation issues, and the fundamental concept remains valid. It appears the CUDAGraph constant buffer modification was unnecessary, so I've removed it.

> if you are trying to speedup the non-cudagraph case mainly and you want to get that done relatively soon, I would suggest to fallback to the old implementation (that launches many kernels) when you detect that cudagraph is being used.

This is a reasonable strategy, but it leads to code bloat, increased binary size, and longer compile time. So I'd prefer getting the approach propose in this PR to work for all cases, as opposed to introducing it as a fast path sooner. Shall we give the direction another shot?

> Introducing a concept of "constant buffer" to cudagraph is going to be a significantly trickier change that I expect is going to take a bit of work to get right.

> the memory will be re-used by other objects within the current cudagraph (since you Tensor object goes out of scope after the kernel ran) and so you cannot rely on the data to be preserved from one run to the next.

Totally agree with the shadiness on introducing the constant buffer concept to `CUDAGraph`. Turns out it's not necessary - as @eellison pointed out, by using `CUDACachingAllocator::raw_alloc`, we reserve a buffer from the graph's private pool. By allowing it to "leak", the buffer won't be reassigned for any other purpose ([reference](https://github.com/pytorch/pytorch/blob/main/c10/cuda/CUDACachingAllocator.cpp#L103-L108)), and its lifetime is tied to the CUDAGraph object. This achieves the desired semantic reliably.

> Moreover, the cudaMemCopy is very shady here, I'm surprised it is actually not raising an error when being captured. It most likely burns in the CPU side pointer and will re-run the copy (from a CPU Tensor that doesn't exist anymore!).

I've confirmed that this copy only occurs once during capture by examining the graph dump:
```
digraph dot {
subgraph cluster_1 {
label="graph_1" graph[style="dashed"];
"graph_1_node_0"[style="bold" shape="octagon" label="0
_ZN2at6native55_GLOBAL__N__791b5f0c_22_ForeachBinaryOpList_cu_a911ec4325multi_tensor_apply_kernelINS1_18TensorListMetadataILi2EEENS1_14UnaryOpFunctorIfLi2ELi1ELi1EEEJNS0_8IdentityIfEEEEEvNS1_12DevArrayPackET0_DpT1_
node handle: 0x00000000086ED680
func handle: 0x00000000085718D0
"];

}
}
```

It not being captured is expected as it was not an async API.

> I'm surprised it is actually not raising

As for this, I believe non-capturable operations doesn't necessarily lead to capture failure. We can guarantee this by temporarily switching to [cudaStreamCaptureModeRelaxed](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__STREAM.html#group__CUDART__STREAM_1g9d0535d93a214cbf126835257b16ba85) via `at::cuda::CUDAStreamCaptureModeGuard`. There's a precedence of this pattern in [CUDACachingAllocator](https://fburl.com/a43ce9ga). It can also be seen from [nccl's cuda graph handling](https://github.com/NVIDIA/nccl/blob/b6475625fbcaa2c3c0e50eed2fa1255d7514d4a2/src/include/alloc.h#L229-L237).

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 3]
```diff
@@ -38,13 +48,184 @@ __device__ __forceinline__ void load_store(
...
+  if (is_capturing) {
+    // The dynamically allocated buffer can be considered an extension of the
+    // kernel argument. Therefore, when used with CUDA graph, the buffer should
+    // remain constant across replays and have the same lifetime as the
+    // captured graph. By allocating the buffer via
+    // CUDACachingAllocator::raw_alloc and allowing it to "leak", a region in
+    // the CUDA graph's pool is reserved for the duration of the graph's
+    // lifetime.
+    pack.buffer_ptr = static_cast<char*>(
+        c10::cuda::CUDACachingAllocator::raw_alloc(total_bytes));
```
Comment (eellison) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: If you do raw_alloc you're responsible for calling `raw_delete` after. Could we just allocate a tensor or storage here ? If we're cudagraphing, there is no C++ overhead so no need to optimize for that.

Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: The intention is not to optimize for CPU overhead, but to mimic the kernel argument semantics of CUDA graph.

If I understand correctly, CUDA graph allocates device memory for kernel arguments and copies them from host to device at capture time. The kernel arguments on device have the same lifetime as the CUDA graph.

Since the high level idea of this PR is to improve `multi_tensor_apply`'s efficiency by not spliting workloads due to the 4kb limitation of the static kernel argument space, the buffer in this context effectively serves as an extension of the kernel argument. Therefore, I thought it might be acceptable to not invoke `raw_delete` here, similar to how regular kernel arguments are never deleted. What do you think?

Comment (eqy) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: Why would a regular ref-counted style allocation not work here? If it's during a graph capture, it should happen on the private memory pool, and while it doesn't look like a huge amount of memory the caching allocator would be free to reclaim it for other purposes once this operation is finished. Note that the requirement is _not_ that the buffer has to be alive for the lifetime of the graph, but rather that each replay can reuse the same address, and this is exactly the use-case that the private memory pool functionality of the caching allocator is supposed to support.

A side question, if this `raw_alloc` happens during a capture on a private pool, would we leak the memory indefinitely, or only until the graph is destroyed?

Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: > Note that the requirement is not that the buffer has to be alive for the lifetime of the graph, but rather that each replay can reuse the same address, and this is exactly the use-case that the private memory pool functionality of the caching allocator is supposed to support.

Thanks for your input, @eqy. In this context, the buffer holds metadata such as pointers and sizes, as opposed to the type of data typically held by tensors. Therefore, I believe it makes more sense to bake the content of the buffer into the graph. The `raw_alloc` combined with capture-time only `cudaMemcpy` aims to achieve this. Consider the following example:

```cpp
constexpr size_t N = ...;
struct Foo {
  int64_t* pointers[N];
}
__global__ my_kernel(Foo foo) {...}
```

During capture, CUDA graph copies the content of `Foo` onto the device, where it remains for the lifetime of the graph. Now, suppose we want to lift the 4KB size limit of `Foo` by replacing pointers with a dynamically allocated buffer (which is effectively the approach used by this PR). I believe it would be reasonable for the buffer to receive the same treatment as `Foo` - reserved for the duration of the graph's lifetime and populated once during capture.

> Why would a regular ref-counted style allocation not work here?

A ref-counted allocation implies that even though we get a baked-in address, the region pointed to by the address could be reassigned for other purposes, leading to the buffer's content being overwritten.

In theory, we could use ref-counted style allocation here and populate the baked-in device location with a HtoD copy on every replay. However, this is unnecessary and inefficient in this context, and I'm uncertain whether HtoD copies can be recorded without direct graph manipulation.

> A side question, if this raw_alloc happens during a capture on a private pool, would we leak the memory indefinitely, or only until the graph is destroyed?

IIUC only until the graph is destroyed.

--------

To summarize, there are two options for this buffer:
- reserve it for the duration of the graph's lifetime and populate it once during capture
- populate its content with a HtoD copy on every replay

I believe the first option makes more sense. Let me know what you think.

Comment (eellison) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: I dont think we are actually skipping the cudaMemcpy currently despite the `raw_alloc`. You can verify with `debug_dump()` in Graph.cpp. 

In either case - we should could a benchmarking demonstrating the win here.

Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: > I dont think we are actually skipping the cudaMemcpy currently despite the raw_alloc. You can verify with debug_dump() in Graph.cpp.

The cudaMemcpy is skipped according to the dump. 

```
> digraph dot {
subgraph cluster_1 {
label="graph_1" graph[style="dashed"];
"graph_1_node_0"[style="bold" shape="octagon" label="0
_ZN2at6native55_GLOBAL__N__791b5f0c_22_ForeachBinaryOpList_cu_a911ec4325multi_tensor_apply_kernelINS1_18TensorListMetadataILi2EEENS1_14UnaryOpFunctorIfLi2ELi1ELi1EEEJNS0_8IdentityIfEEEEEvNS1_12DevArrayPackET0_DpT1_
node handle: 0x00000000086ED680
func handle: 0x00000000085718D0
"];

}
}
```

Also, according to https://www.nvidia.com/en-us/on-demand/session/gtcspring23-s51211/, the non-async version of cudaMemcpy is not capturable.

image


> In either case - we should could a benchmarking demonstrating the win here.

I'm not sure whether the alternative, i.e. recording a HtoD copy can be achievied via stream capture (no direct graph manipulation). Do you know?

Comment (eqy) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: In that case, what prevents us from giving the optimizer a (non-parameter) tensor member (of e.g., dtype `bytes`), that is allocated on the default memory pool and only created when e.g., a `use_graphs` argument is set to `True` similar to `capturable` here: https://github.com/pytorch/pytorch/blob/c802c5019602d5ae6e43f8079d6bbad48a53ef60/torch/optim/adam.py#L25 ?

Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: @eqy thanks for offering the idea. It is indeed a good pattern for this type of problems in general.

However, the main challenge of applying it here is that not all upstream users of `multi_tensor_apply` are stateful like optimizers. `foreach` kernels are also used without optimizers in many cases. Requring users to maintain and pass in a "workspace" buffer to make any `foreach` op capturable seems like a degradation in UX.

We can give this approach more thought. However, do you see any substantial downsides to the proposed approach, aside from it being a bit unconventional? It only appears that we are leaking memory, but this is not actually the case since the lifetime of the allocation is tied to the graph.

Comment (eqy) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:181]: I think the reasons I am giving are purely aesthetic, mainly that this extra buffer seems to be a special case of multi-tensor apply, and intuitively the "special-case" handling might make more sense to have in the thing that actually is the special case rather than the generic CUDA graphs code.

I agree this is justifiable to enable e.g., the non-optimizer use cases of `multi_tensor_apply` as you mentioned.

**Comment by yifuwang** [comment]
I had an offline discussion with @eellison. He brought up a good point that a memory pool can be shared by multiple graphs created via `make_graphed_callables`. While this may not render the leaking-in-the-private-pool solution incorrect, it certainly makes it less appealing, as the allocation's lifetime would be tied to multiple graphs.

We agreed that the best way to approach the problem is **to first replicate the non-CUDA graph behavior (i.e. copying the args to device on every replay) before optimizing upon it**. Even in this scenario, we would need a host buffer with the same lifetime as the graph, populated upon capturing. This is because the necessary information, such as tensor points/sizes, won't be available during replays.

After some digging, I found that [CUDA User Objects](https://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#cuda-user-objects) offers exactly what we need. The combination of [cudaUserObjectCreate](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1g0aa7caf7bb7ccbc766fab0558af72c96) and [cudaGraphRetainUserObject](https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__GRAPH.html#group__CUDART__GRAPH_1g95b48aa4255fe6803cd37f8e0c81c1fa) allows us to transfer the ownership of a buffer to the currently capturing graph.

I've update the PR so it works as follows:
- During cuda graph capture, allocate a host buffer to store the extended launch parameters
- Create a CUDA User Object to manage the lifetime of this buffer
- Transfer the ownership of the CUDA User Object to the currently capturing graph
- Allocate/free the device buffer the for the extended launch parameters as usual
- Capture the HtoD copy as part of the CUDA Graph (verified via `debug_dump`)

@albanD, @eellison, @eqy - would love to hear your thoughts on this approach. If this looks reasonable, we can consider a CUDA User Object-managed device buffer to avoid the HtoD copies during replays.

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 3 above)
Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh]: I echo your point on aesthetics. I updated the PR to use CUDA User Object to manage the lifetime of the extended launch arguments. Hopefully it addresses the concern.

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 4]
```diff
@@ -38,13 +49,213 @@ __device__ __forceinline__ void load_store(
...
+    const Vectors&... vecs) {
+  std::vector<const void*> ptrs;
+  std::vector<size_t> sizes;
+  std::vector<size_t> offsets;
+  size_t total_bytes = 0;
+
+  for (auto d = 0; d < n; ++d) {
+    pack_vectors_helper(ptrs, sizes, offsets, total_bytes, addresses[d]);
+  }
+  (pack_vectors_helper(ptrs, sizes, offsets, total_bytes, vecs), ...);
```
Comment (albanD) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:144]: What is the `, ...)` for here?

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 5]
```diff
@@ -38,13 +49,213 @@ __device__ __forceinline__ void load_store(
...
+    // issued on every replay. This copy can be eliminated by having the CUDA
+    // User Object manage a device buffer.
+    uint8_t* buf = new uint8_t[total_bytes];
+
+    // Manage the ownership of buf with a cuda user object
+    cudaUserObject_t user_object;
+    C10_CUDA_CHECK(cudaUserObjectCreate(
+        &user_object,
+        buf,
+        [](void* buf) { delete[] reinterpret_cast<uint8_t*>(buf); },
```
Comment (albanD) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:190]: The raw "new" is not the best... Can you not capture the Tensor by value in this lambda to keep it alive until the lambda is called and them reset the shared_ptr in the lambda to free it? :D

**Review by albanD — COMMENTED** [review]
Thanks for the update. That sounds pretty cool!

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 4 above)
Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:144]: This is a fold expression. It calls `pack_vectors_helper` on every element in the parameter pack `vecs`. The expression allows `pack_vectors_helper` to be specialized based on the type of each element in the parameter pack. Without this, we would have to type-erase the elements in `vecs` to make this function reusable for other MTA flavors.

Comment (albanD) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:144]: if that C++ 17 feature is supported by all the compilers we have in CI (MSVC in particular), I'm happy with it :D

Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:144]: Ah good point. Probably not. Will probably need to replace this with some template-based expansion for now.

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 6]
```diff
@@ -23,6 +32,8 @@ static constexpr int depth_to_max_tensors_scalarlist_of_complex_double[2] = {
     72,
     60};
 
+static constexpr size_t max_kernel_arg_size = 4096;
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:35]: just fyi (I don't mean we should update this value at all in this pr)
> CUDA 12.1 increases this parameter limit from 4,096 bytes to 32,764 bytes on all device architectures including [NVIDIA Volta](https://www.nvidia.com/en-us/data-center/volta-gpu-architecture/) and above.

ref: https://developer.nvidia.com/blog/cuda-12-1-supports-large-kernel-parameters/

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 7]
```diff
@@ -38,13 +49,213 @@ __device__ __forceinline__ void load_store(
...
+//
+// However, when the problem size is small, the latency of the cudaMemcpyAsync
+// becomes more pronounced. In such cases, we try to fit the arrays into the
+// 4kb kernel argument space. The technique is akin to small buffer
+// optimization.
+//
+// However, when the problem size is small, the latency of the cudaMemcpyAsync
+// becomes more pronounced. In such cases, we attempt to fit the arrays into
+// the 4KB kernel argument space. This technique is akin to the small buffer
+// optimization.
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:71]: looks duplicated
```suggestion
```

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 8]
```diff
@@ -38,13 +49,213 @@ __device__ __forceinline__ void load_store(
...
+  if (total_bytes < DevArrayPack::small_buffer_size) {
+    pack.buffer_ptr = nullptr;
+    for (const auto i : c10::irange(ptrs.size())) {
+      pack.offsets[i] = offsets[i];
+      memcpy(pack.small_buffer + offsets[i], ptrs[i], sizes[i]);
+    }
+    return std::make_tuple(pack, c10::optional<at::Tensor>(c10::nullopt));
+  }
+
+  bool is_capturing = at::cuda::currentStreamCaptureStatusMayInitCtx() !=
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:159]: nit-picking
```suggestion
  const bool is_capturing = at::cuda::currentStreamCaptureStatusMayInitCtx() !=
```

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 9]
```diff
@@ -38,13 +49,213 @@ __device__ __forceinline__ void load_store(
...
+        // Only use pinned memory when not capturing
+        at::TensorOptions().dtype(at::kByte).pinned_memory(true));
+  } else {
+    // With CUDA Graph, we need the data to have the same lifetime as the
+    // capturing graph because the data is an extension of the launch argument.
+    // This is achieved by dynamically allocating a buffer, managing the
+    // lifetime of the buffer with a CUDA User Object, and transferring the
+    // ownership of the CUDA User Object to the capturing graph.
+    //
+    // NOTE: based on PR feedback, we want to be conservative and replicate the
```
Comment (crcrpar) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:179]: would it be possible to put the link to this feedback here?

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 4 above)
Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh]: Thanks for raising the point about compiler compatibility. I've replaced this with a template-based expansion

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 6 above)
Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:34]: This is awesome!! Thank you for bringing this up. We should adjust this based-on device generation and driver version soon.

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh (hunk 5 above)
Comment (yifuwang) [reply aten/src/ATen/native/cuda/MultiTensorApply.cuh:200]: Yeah agreed. However, `cudaUserObjectCreate` takes a raw function pointer of type `void(CUDART_CB* cudaHostFn_t )( void*  userData )` which means capturing a tensor ref isn't possible.

**Review by eellison — COMMENTED** [review]
Fix test failure or rebase if not related ?

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 10]
```diff
@@ -38,13 +48,224 @@ __device__ __forceinline__ void load_store(
...
+        &capture_status,
+        nullptr, // id_out
+        &graph));
+
+    // Transfer the ownership to the graph
+    // NOTE: according to the documentation of cudaStreamGetCaptureInfo, all
+    // operations other than destroy and node removal are permitted on the
+    // graph while the capture sequence is in progress.
+    C10_CUDA_CHECK(cudaGraphRetainUserObject(
+        graph, user_object, 1, cudaGraphUserObjectMove));
```
Comment (eqy) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:218]: probably need to add an entry to https://github.com/pytorch/pytorch/blob/main/torch/utils/hipify/cuda_to_hip_mappings.py to make HIP build happy

File: aten/src/ATen/native/cuda/MultiTensorApply.cuh [hunk 11]
```diff
@@ -204,7 +204,7 @@ std::tuple<DevArrayPack, c10::optional<at::Tensor>> pack_vectors(
     // Query the currently capturing graph
     cudaStreamCaptureStatus capture_status;
     cudaGraph_t graph;
-    C10_CUDA_CHECK(cudaStreamGetCaptureInfo(
+    C10_CUDA_CHECK(cudaStreamGetCaptureInfo_v2(
```
Comment (eqy) [inline aten/src/ATen/native/cuda/MultiTensorApply.cuh:207]: This probably also needs to be added to the `cuda_to_hip_mappings` ;;

**Comment by jeffdaily** [comment]
Looks like also missing cuda_to_hip_mappings for 
- cudaGraphUserObjectMove --> hipGraphUserObjectMove
- cudaUserObjectNoDestructorSync --> hipUserObjectNoDestructorSync
- cudaUserObject_t --> hipUserObject_t

**Comment by yifuwang** [comment]
Thanks for everyone's help and inputs (and I apologize for the delay). I've updated the PR:
- During capture, instead of using a `cudaUserObject_t` to manage a copy of the launch params in RAM and issue an HtoD copy on every reply, we use the `cudaUserObject_t` to manage a copy of the launch params on device. I've verified that memory gets freed as expected upon graph destruction.
- Added a unit test to cover multi_tensor_apply + cuda graph + dynamically allocated launch param.

Updated synthetic benchmark (foreach copying 1500 tensors, totaling about 0.64GB) on an H100 /w HBM2E (sorry the links are Meta internal):


Baseline without cuda graph:
- memory bandwidth: 1472.627 GB/s
- https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_23c9373e-87e6-4362-b4e1-02a6c8112dbe.json

Baseline with cuda graph:
- memory bandwidth: 1528.603 GB/s
- https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_ab27dbb9-2057-458d-b4cd-da7b27db8430.json

This PR without cuda graph:
- memory bandwidth: 1923.237 GB/s
- https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_c843bd98-42df-4247-b3fd-95afde0dca2e.json

This PR with cuda graph:
- memory bandwidth: 1948.886 GB/s
- https://interncache-all.fbcdn.net/manifold/perfetto-artifacts/tree/ui/index.html#!/?url=https://interncache-all.fbcdn.net/manifold/perfetto_internal_traces/tree/shared_trace/yifu_848e2612-7f73-4180-8b5f-601e2cadae7b.json

**Review by eqy — APPROVED** [review]

**Review by eellison — APPROVED** [review]

**Review by crcrpar — APPROVED** [review]

**Comment by yifuwang** [comment]
Fixed the issue with ROCm.

The problem was that `{cuda,hip}StreamCaptureModeRelaxed` behaves differently in CUDA and HIP - `cudaMemcpy` is permitted in CUDA under relaxed capture mode, but not in HIP. After some investigation, I found that the combination of `cudaMemcpyAsync` + `cudaStreamSynchronize` is compatible with both CUDA and HIP under relaxed capture mode. Conveniently, this combination exists as `at::cuda::memcpy_and_sync`.

With the issue resolved, I had the opportunity to benchmark the optimization on MI300X. However, I found that it led to a significant regression. After further investigation, I discovered that the HIP compiler wasn't able to generate warp-sync code for an 'if' statement in the small buffer optimization logic, whereas `nvcc` could. I managed to fix the regression by refactoring the code so that the logic now only deals with constexpr.

Attached the numbers for `foreach_copy`ing the following weights on an `MI300x`:

```python
SIZES = [
    torch.Size([8192, 8192]),
    torch.Size([1024, 8192]),
    torch.Size([1024, 8192]),
    torch.Size([8192, 8192]),
    torch.Size([28672, 8192]),
    torch.Size([8192, 28672]),
    torch.Size([28672, 8192]),
    torch.Size([8192]),
    torch.Size([8192]),
]
```

Baseline:
```
gb: 3.18756103515625
device ms: 2.384, cpu ms: 0.154
memory bandwidth: 2674.263 GB/s
```

This PR:
```
gb: 3.18756103515625
device ms: 2.077, cpu ms: 0.053
memory bandwidth: 3068.742 GB/s
```

cc @xw285cornell
